app/src/ingest_edd_web.py

```diff
@@ -20,55 +22,133 @@ def _ingest_edd_web(
+        logger.info("Processing: %s (%s)", name, item["url"])
         urls_processed.add(item["url"])
 
         content = item.get("main_content", item.get("main_primary"))
         assert content, f"Item {name} has no main_content or main_primary"
 
         document = Document(name=name, content=content, source=item["url"], **doc_attribs)
-        db_session.add(document)
-
-        chunks = _chunk_page(content)
-        for chunk in chunks:
-            chunk.document = document
-        db_session.add_all(chunks)
-        logger.info("Split %s into %d chunks", name, len(chunks))
+        chunks, splits = _chunk_page(document, content)
+        result.append((document, chunks, splits))
+    return result
 
 
-def _chunk_page(content: str) -> list[Chunk]:
-    embedding_model = app_config.sentence_transformer
+def _chunk_page(
+    document: Document, content: str
+) -> tuple[Sequence[Chunk], Sequence[SplitWithContextText]]:
+    splits: list[SplitWithContextText] = []
+    for headings, text in split_markdown_by_heading(f"# {document.name}\n\n" + content):
+        # Start a new split for each heading
+        section_splits = _split_heading_section(headings, text)
+        splits.extend(section_splits)
 
-    # Split content by double newlines, then gather into the largest chunks
-    # that tokenize to less than the max_seq_length
-    content_split_by_double_newlines = content.split("\n\n")
-    subsections = [content_split_by_double_newlines[0]]
-    for subsection in content_split_by_double_newlines[1:]:
-        tokens_with_new_subsection = len(tokenize(subsections[-1] + subsection))
-        if tokens_with_new_subsection < embedding_model.max_seq_length:
-            subsections[-1] += "\n\n" + subsection
-        else:
-            subsections.append(subsection)
-
-    # Parallelize embedding generation for performance
-    subsection_embeddings = embedding_model.encode(subsections, show_progress_bar=False)
-
-    return [
+    chunks = [
         Chunk(
-            content=subsection,
-            mpnet_embedding=embedding,
-            tokens=len(tokenize(subsection)),
+            document=document,
+            content=split.text,
+            headings=split.headings,
+            num_splits=len(splits),
+            split_index=index,
+            tokens=split.token_count,
         )
-        for subsection, embedding in zip(subsections, subsection_embeddings, strict=True)
+        for index, split in enumerate(splits)
     ]
+    return chunks, splits
+
+
+# MarkdownHeaderTextSplitter splits text by "\n" then calls aggregate_lines_to_chunks() to reaggregate
+# paragraphs using "  \n", so "\n\n" is replaced by "  \n"
+MarkdownHeaderTextSplitter_DELIMITER = "  \n"
+
+
+def _split_heading_section(headings: Sequence[str], text: str) -> list[SplitWithContextText]:
+    # Add headings to the context_str; other context can also be added
+    context_str = headings_as_markdown(headings)
```

Comment:
A stray though, but if `context_str` is only ever intended for generating the embedding, we may not need to render these to markdown (since I don't think the markdown would have any impact on search besides generating additional tokens in the embedding)

Reply:
I rendered it as markdown because many LLM an interpret markdown. I'm assuming it distinguishes heading text from regular text and puts some more significance on the heading text -- I could be totally wrong, but I played it safe and added the few extra `#`s for headings.

Reply:
But at the moment `context_str` is only going to `SplitWithContextText.text_to_encode`, i.e., it'll never be seen by the LLM, right?

Reply:
Oh right! I got it mixed up. Want me to replace it with this?
```suggestion
    context_str = "\n".join(headings)
```

Reply:
That makes sense to me! Might save a few tokens.

Reply:
Replaced.